ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java

```diff
@@ -2680,9 +2683,9 @@ public void run() {
                   total += estimator.estimate(jobConf, scanOp, -1).getTotalLength();
                 }
                 recordSummary(path, new ContentSummary(total, -1, -1));
-              } else {
-                // todo: should nullify summary for non-native tables,
-                // not to be selected as a mapjoin target
+              } else if (handler == null) {
```

Comment:
This seems unrelated for me. Maybe a different PR?

Reply:
Thank you @pvary.  The pr originally is trying to solve the non-native table that participates in `join`, such table will be created with a path, but have no data under it, so the non-native table may be selected as a mapjoin target, which should cause oom problem when building the hashtable. The changes take place in one file, so I put them together.

Reply:
What happened before this change?

Reply:
Could you please help explain?

Reply:
Last two months, in our cluster we found that join a kudu table(mr engine) cloud make the hiveserver2 oom.  [HMSHandler](https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java#L2385-L2407) seems create a path for the non-native table, so `fs.getContentSummary(path)` returns the file length with 0(if data does not store under the table dir like kudu, hbase), the optimizer will [collect](https://github.com/apache/hive/blob/7b3ecf617a6d46f48a3b6f77e0339fd4ad95a420/ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/AbstractJoinTaskDispatcher.java#L118) the sizes of join tables before converting the join to mapjoin(by default), if the non-native table is able to be a stream table(like inner join) and the length of `his data` is 0,  the non-native will be tagged as small table and a local task will be started to build the hashtable. If the non-native table has millions of records, the jvm may not be able to hold all these records in memory.

Reply:
If the handler is not null (we have some specific handler which is not an `InputEstimator`), then we do not add the file sizes of data size, but we still have a `totalLength`, `totalFileCount`, `totalDirectoryCount` as 0?
Do I miss something?

Reply:
The summary will be recorded in current [Context](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L2709) and the change maps the non-native path to null as it ignores [recording summary](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L2700). The [Optimizer](https://github.com/apache/hive/blob/7b3ecf617a6d46f48a3b6f77e0339fd4ad95a420/ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/AbstractJoinTaskDispatcher.java#L133) retrieves the summary from the same Context instance, as it is null, indicates that the table is not a good choice for [mapjoin target](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L2757-L2760).

Reply:
Thanks for the explanation and the patience! 😄

Reply:
😄 Thank you for the review!